File: python/src/metadata_ai/agent.py
"""Agent handle for the Metadata AI SDK."""

from __future__ import annotations

from collections.abc import AsyncIterator, Iterable
from typing import Any

from metadata_ai._http import AsyncHTTPClient, HTTPClient
from metadata_ai._logging import debug as _log_debug
from metadata_ai._streaming import AsyncSSEIterator, SSEIterator
from metadata_ai.models import AgentInfo, InvokeRequest, InvokeResponse, StreamEvent


def _debug(msg: str) -> None:
    """Print debug message if debug is enabled."""
    _log_debug("AGENT DEBUG", msg)


class AgentHandle:
    """
    Handle for invoking a specific agent.

    Usage:
        agent = client.agent("DataQualityPlannerAgent")

        # Synchronous call
        response = agent.call("What tables have issues?")
        print(response.response)

        # Streaming call
        for event in agent.stream("What tables have issues?"):
            if event.type == "content":
                print(event.content, end="", flush=True)
    """

    def __init__(
        self,
        name: str,
        http: HTTPClient,
        async_http: AsyncHTTPClient | None = None,
    ):
        """
        Initialize the agent handle.

        Args:
            name: The agent name
            http: HTTP client for API communication
            async_http: Optional async HTTP client for async operations
        """
        self._name = name
        self._http = http
        self._async_http = async_http

    @property
    def name(self) -> str:
        """Get the agent name."""
        return self._name

    def call(
        self,
        message: str | None = None,
        *,
        conversation_id: str | None = None,
        parameters: dict[str, Any] | None = None,
    ) -> InvokeResponse:
        """
        Invoke the agent synchronously.

        Args:
            message: The query or instruction for the agent. Optional if the agent
                has a default prompt configured.
            conversation_id: Optional ID for multi-turn conversations
            parameters: Optional parameters to pass to the agent

        Returns:
            InvokeResponse with the complete response

        Raises:
            AgentNotFoundError: If the agent does not exist
            AgentNotEnabledError: If the agent is not API-enabled
            AuthenticationError: If the token is invalid
            AgentExecutionError: If the agent execution fails
        """
        _debug("=== AgentHandle.call() START ===")
        _debug(f"Agent: {self._name}")
        _debug(f"Message: {message}")
        _debug(f"Conversation ID: {conversation_id}")

        request = InvokeRequest(
            message=message,
            conversation_id=conversation_id,
            parameters=parameters or {},
        )
        _debug(f"Request dict: {request.to_api_dict()}")

        data = self._http.post(
            f"/{self._name}/invoke",
            json=request.to_api_dict(),
            agent_name=self._name,
        )

        _debug(f"Raw API response data: {data}")

        response = InvokeResponse.from_dict(data)
        _debug(f"Parsed response.response: {response.response}")
        _debug(f"Parsed response.tools_used: {response.tools_used}")
        _debug("=== AgentHandle.call() END ===")

        return response

    async def acall(
        self,
        message: str | None = None,
        *,
        conversation_id: str | None = None,
        parameters: dict[str, Any] | None = None,
    ) -> InvokeResponse:
        """
        Invoke the agent asynchronously.

        Args:
            message: The query or instruction for the agent. Optional if the agent
                has a default prompt configured.
            conversation_id: Optional ID for multi-turn conversations
            parameters: Optional parameters to pass to the agent

        Returns:
            InvokeResponse with the complete response

        Raises:
            AgentNotFoundError: If the agent does not exist
            AgentNotEnabledError: If the agent is not API-enabled
            AuthenticationError: If the token is invalid
            AgentExecutionError: If the agent execution fails
            RuntimeError: If async client is not available
        """
        if self._async_http is None:
            raise RuntimeError(
                "Async HTTP client not available. "
                "Use MetadataAI with enable_async=True for async operations."
            )

        request = InvokeRequest(
            message=message,
            conversation_id=conversation_id,
            parameters=parameters or {},
        )
        data = await self._async_http.post(
            f"/{self._name}/invoke",
            json=request.to_api_dict(),
            agent_name=self._name,
        )
        return InvokeResponse.from_dict(data)

    def stream(
        self,
        message: str | None = None,
        *,
        conversation_id: str | None = None,
        parameters: dict[str, Any] | None = None,
    ) -> Iterable[StreamEvent]:
        """
        Invoke the agent with streaming response.

        Args:
            message: The query or instruction for the agent. Optional if the agent
                has a default prompt configured.
            conversation_id: Optional ID for multi-turn conversations
            parameters: Optional parameters to pass to the agent

        Yields:
            StreamEvent objects as they arrive

        Raises:
            AgentNotFoundError: If the agent does not exist
            AgentNotEnabledError: If the agent is not API-enabled
            AuthenticationError: If the token is invalid
            AgentExecutionError: If the agent execution fails
        """
        request = InvokeRequest(
            message=message,
            conversation_id=conversation_id,
            parameters=parameters or {},
        )
        byte_stream = self._http.post_stream(
            f"/{self._name}/stream",
            json=request.to_api_dict(),
            agent_name=self._name,
        )
        return SSEIterator(byte_stream)

    def astream(
        self,
        message: str | None = None,
        *,
        conversation_id: str | None = None,
        parameters: dict[str, Any] | None = None,
    ) -> AsyncIterator[StreamEvent]:
        """
        Invoke the agent with async streaming response.

        Args:
            message: The query or instruction for the agent. Optional if the agent
                has a default prompt configured.
            conversation_id: Optional ID for multi-turn conversations
            parameters: Optional parameters to pass to the agent

        Yields:
            StreamEvent objects as they arrive

        Raises:
            AgentNotFoundError: If the agent does not exist
            AgentNotEnabledError: If the agent is not API-enabled
            AuthenticationError: If the token is invalid
            AgentExecutionError: If the agent execution fails
            RuntimeError: If async client is not available
        """
        if self._async_http is None:
            raise RuntimeError(
                "Async HTTP client not available. "
                "Use MetadataAI with enable_async=True for async operations."
            )

        request = InvokeRequest(
            message=message,
            conversation_id=conversation_id,
            parameters=parameters or {},
        )
        byte_stream = self._async_http.post_stream(
            f"/{self._name}/stream",
            json=request.to_api_dict(),
            agent_name=self._name,
        )
        return AsyncSSEIterator(byte_stream)

    def get_info(self) -> AgentInfo:
        """
        Get agent metadata including abilities and description.

        Returns:
            AgentInfo with agent metadata

        Raises:
            AgentNotFoundError: If the agent does not exist
            AgentNotEnabledError: If the agent is not API-enabled
        """
        data = self._http.get(f"/{self._name}", agent_name=self._name)
        return AgentInfo.from_dict(data)

    async def aget_info(self) -> AgentInfo:
        """
        Get agent metadata asynchronously.

        Returns:
            AgentInfo with agent metadata

        Raises:
            AgentNotFoundError: If the agent does not exist
            AgentNotEnabledError: If the agent is not API-enabled
            RuntimeError: If async client is not available
        """
        if self._async_http is None:
            raise RuntimeError(
                "Async HTTP client not available. "
                "Use MetadataAI with enable_async=True for async operations."
            )

        data = await self._async_http.get(f"/{self._name}", agent_name=self._name)
        return AgentInfo.from_dict(data)

    def __repr__(self) -> str:
        return f"AgentHandle(name={self._name!r})"

File: typescript/src/agent.ts
/**
 * Agent handle for the Metadata AI SDK.
 *
 * This module provides the AgentHandle class for invoking specific agents.
 */

import type { HttpClient } from './http.js';
import {
  mapAgentInfo,
  type AgentInfo,
  type ApiAgentInfo,
  type ApiInvokeResponse,
  type InvokeOptions,
  type InvokeRequestBody,
  type InvokeResponse,
  type StreamEvent,
} from './models.js';
import { createStreamIterable } from './streaming.js';

/**
 * Convert API invoke response to InvokeResponse.
 */
function mapInvokeResponse(data: ApiInvokeResponse): InvokeResponse {
  return {
    conversationId: data.conversationId,
    response: data.response,
    toolsUsed: data.toolsUsed || [],
    usage: data.usage
      ? {
          promptTokens: data.usage.promptTokens || 0,
          completionTokens: data.usage.completionTokens || 0,
          totalTokens: data.usage.totalTokens || 0,
        }
      : undefined,
  };
}

/**
 * Handle for invoking a specific agent.
 *
 * This class provides methods for synchronous and streaming invocation
 * of Metadata AI agents.
 *
 * @example
 * ```typescript
 * // Get agent handle from client
 * const agent = client.agent('DataQualityPlannerAgent');
 *
 * // Synchronous invocation
 * const response = await agent.invoke('What tables have issues?');
 * console.log(response.response);
 *
 * // Streaming invocation
 * for await (const event of agent.stream('Analyze the orders table')) {
 *   if (event.type === 'content') {
 *     process.stdout.write(event.content || '');
 *   }
 * }
 * ```
 */
export class AgentHandle {
  private readonly agentName: string;
  private readonly http: HttpClient;

  /**
   * Create a new agent handle.
   *
   * @param name - The agent name
   * @param http - HTTP client for API communication
   * @internal This constructor is called by AiSdk.agent()
   */
  constructor(name: string, http: HttpClient) {
    this.agentName = name;
    this.http = http;
  }

  /**
   * Get the agent name.
   */
  get name(): string {
    return this.agentName;
  }

  /**
   * Invoke the agent synchronously.
   *
   * @param message - Optional query or instruction for the agent. If not provided, the agent's configured prompt will be used.
   * @param options - Optional invoke options (conversationId, parameters)
   * @returns Promise resolving to the complete response
   *
   * @throws {AgentNotFoundError} If the agent does not exist
   * @throws {AgentNotEnabledError} If the agent is not API-enabled
   * @throws {AuthenticationError} If the token is invalid
   * @throws {AgentExecutionError} If the agent execution fails
   *
   * @example
   * ```typescript
   * // Simple invocation
   * const response = await agent.invoke('What tables have quality issues?');
   * console.log(response.response);
   * console.log('Tools used:', response.toolsUsed);
   *
   * // Invoke without message (uses agent's configured prompt)
   * const response = await agent.invoke();
   *
   * // Multi-turn conversation
   * const r1 = await agent.invoke('Analyze the orders table');
   * const r2 = await agent.invoke('Now create tests for the issues', {
   *   conversationId: r1.conversationId,
   * });
   * ```
   */
  async invoke(message?: string, options?: InvokeOptions): Promise<InvokeResponse> {
    const requestBody: InvokeRequestBody = {};

    if (message !== undefined) {
      requestBody.message = message;
    }

    if (options?.conversationId) {
      requestBody.conversationId = options.conversationId;
    }

    if (options?.parameters && Object.keys(options.parameters).length > 0) {
      requestBody.parameters = options.parameters;
    }

    const data = await this.http.post<ApiInvokeResponse>(
      `/name/${this.agentName}/invoke`,
      requestBody,
      this.agentName
    );

    return mapInvokeResponse(data);
  }

  /**
   * Invoke the agent with streaming response.
   *
   * @param message - Optional query or instruction for the agent. If not provided, the agent's configured prompt will be used.
   * @param options - Optional invoke options (conversationId, parameters)
   * @returns Async iterable of stream events
   *
   * @throws {AgentNotFoundError} If the agent does not exist
   * @throws {AgentNotEnabledError} If the agent is not API-enabled
   * @throws {AuthenticationError} If the token is invalid
   * @throws {AgentExecutionError} If the agent execution fails
   *
   * @example
   * ```typescript
   * for await (const event of agent.stream('Analyze data quality')) {
   *   switch (event.type) {
   *     case 'start':
   *       console.log('Started conversation:', event.conversationId);
   *       break;
   *     case 'content':
   *       process.stdout.write(event.content || '');
   *       break;
   *     case 'tool_use':
   *       console.log('Using tool:', event.toolName);
   *       break;
   *     case 'end':
   *       console.log('\nCompleted');
   *       break;
   *     case 'error':
   *       console.error('Error:', event.error);
   *       break;
   *   }
   * }
   *
   * // Stream without message (uses agent's configured prompt)
   * for await (const event of agent.stream()) {
   *   // handle events
   * }
   * ```
   */
  async *stream(
    message?: string,
    options?: InvokeOptions
  ): AsyncGenerator<StreamEvent, void, unknown> {
    const requestBody: InvokeRequestBody = {};

    if (message !== undefined) {
      requestBody.message = message;
    }

    if (options?.conversationId) {
      requestBody.conversationId = options.conversationId;
    }

    if (options?.parameters && Object.keys(options.parameters).length > 0) {
      requestBody.parameters = options.parameters;
    }

    const byteStream = await this.http.postStream(
      `/name/${this.agentName}/stream`,
      requestBody,
      this.agentName
    );

    yield* createStreamIterable(byteStream);
  }

  /**
   * Get agent metadata including abilities and description.
   *
   * @returns Promise resolving to agent information
   *
   * @throws {AgentNotFoundError} If the agent does not exist
   * @throws {AgentNotEnabledError} If the agent is not API-enabled
   *
   * @example
   * ```typescript
   * const info = await agent.getInfo();
   * console.log('Agent:', info.displayName);
   * console.log('Description:', info.description);
   * console.log('Abilities:', info.abilities.join(', '));
   * ```
   */
  async getInfo(): Promise<AgentInfo> {
    const data = await this.http.get<ApiAgentInfo>(
      `/name/${this.agentName}`,
      undefined,
      this.agentName
    );

    return mapAgentInfo(data);
  }

  /**
   * String representation of the agent handle.
   */
  toString(): string {
    return `AgentHandle(name="${this.agentName}")`;
  }
}
